// Copyright (c) 2020 The Orbit Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";

package orbit_grpc_protos;

import "module.proto";
import "tracepoint.proto";

// This is needed by LockFreeBufferCaptureEventProducer.
option cc_enable_arenas = true;

message InstrumentedFunction {
  reserved 4;

  string file_path = 1;
  string file_build_id = 6;
  uint64 file_offset = 2;
  uint64 function_id = 3;
  uint64 function_size = 7;
  string function_name = 5;

  bool record_arguments = 8;
  bool record_return_value = 9;
}

// Api functions are declared in Orbit.h. They are implemented in user code
// through the ORBIT_API_INSTANTIATE macro. Those functions are used to query
// the tracee for Orbit specific information. See "FindApiFunction" for more
// details.
message ApiFunction {
  string module_path = 1;
  string module_build_id = 2;
  uint64 address = 3;
  string name = 4;
  uint64 api_version = 5;
}

// NextId: 19
message CaptureOptions {
  reserved 17;

  bool trace_context_switches = 1;
  uint32 pid = 2;
  double samples_per_second = 3;
  // Expected to be "uint16".
  uint32 stack_dump_size = 16;

  enum UnwindingMethod {
    kUndefined = 0;
    kFramePointers = 1;
    kDwarf = 2;
  }
  UnwindingMethod unwinding_method = 4;

  enum DynamicInstrumentationMethod {
    kDynamicInstrumentationMethodUnspecified = 0;
    kKernelUprobes = 1;
    kUserSpaceInstrumentation = 2;
  }
  DynamicInstrumentationMethod dynamic_instrumentation_method = 18;

  repeated InstrumentedFunction instrumented_functions = 5;

  bool trace_thread_state = 8;

  bool trace_gpu_driver = 6;

  repeated TracepointInfo instrumented_tracepoint = 7;

  bool enable_introspection = 9;

  uint64 max_local_marker_depth_per_command_buffer = 10;

  bool collect_memory_info = 11;
  uint64 memory_sampling_period_ns = 12;

  repeated ApiFunction api_functions = 13;

  bool enable_api = 14;
}

// For CaptureEvents with a duration, excluding for now GPU-related ones, we
// represent the time interval as (duration_ns, end_timestamp_ns). As durations
// are smaller in magnitude than absolute timestamps, and as protobufs use
// variable-length encoding, this saves some bandwidth compared to
// (start_timestamp_ns, end_timestamp_ns).
// The pair (duration_ns, end_timestamp_ns) was preferred over
// (start_timestamp_ns, duration_ns) because this way, when perf_event_open
// events are processed in order, the CaptureEvents are sorted by the only
// absolute timestamp contained in them (as they are obviously generated when
// the "end" perf_event_open event is processed).

message SchedulingSlice {
  reserved 4;
  uint32 pid = 1;
  uint32 tid = 2;
  int32 core = 3;
  uint64 duration_ns = 6;
  uint64 out_timestamp_ns = 5;
}

message FunctionCall {
  reserved 4;
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 function_id = 3;
  uint64 duration_ns = 9;
  uint64 end_timestamp_ns = 5;
  int32 depth = 6;
  uint64 return_value = 7;
  repeated uint64 registers = 8;
}

// FunctionEntry and FunctionExit are emitted by user space instrumentation.
message FunctionEntry {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 function_id = 3;
  uint64 stack_pointer = 4;
  uint64 return_address = 5;
  uint64 timestamp_ns = 6;
}

message FunctionExit {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;
}

message ApiEvent {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  // The fields below are used to encode orbit_api::EncodedEvent events. This is
  // done mainly for the performance gain of not having to allocate memory out
  // of the Arena for strings, but also to emulate the old uprobe-based use of
  // registers to encode events. To review after deprecation (b/182485096).
  fixed64 r0 = 4;
  fixed64 r1 = 5;
  fixed64 r2 = 6;
  fixed64 r3 = 7;
  fixed64 r4 = 8;
  fixed64 r5 = 9;
}

message ApiScopeStart {
  // NextID: 16

  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  // We encode the first 64 characters of the scope's name/label using fixed64
  // values, in order to avoid an additional allocation (for the internal
  // vector) for short strings. Those allocations have shown a huge performance
  // impact.
  // If the name exceeds the 64 characters, the `encoded_name_additional` will
  // be used to encode the additional characters. This is also a sequence of
  // `fixed64` instead of a `bytes` field so that it can be allocated on an
  // `Arena` (`std::string`s cannot be allocated in `Arena`s).
  fixed64 encoded_name_1 = 4;
  fixed64 encoded_name_2 = 5;
  fixed64 encoded_name_3 = 6;
  fixed64 encoded_name_4 = 7;
  fixed64 encoded_name_5 = 8;
  fixed64 encoded_name_6 = 9;
  fixed64 encoded_name_7 = 10;
  fixed64 encoded_name_8 = 11;
  repeated fixed64 encoded_name_additional = 12;

  uint32 color_rgba = 13;
  uint64 group_id = 14;
  uint64 address_in_function = 15;
}

message ApiScopeStop {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;
}

message ApiScopeStartAsync {
  // NextID: 16

  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 4;
  fixed64 encoded_name_2 = 5;
  fixed64 encoded_name_3 = 6;
  fixed64 encoded_name_4 = 7;
  fixed64 encoded_name_5 = 8;
  fixed64 encoded_name_6 = 9;
  fixed64 encoded_name_7 = 10;
  fixed64 encoded_name_8 = 11;
  repeated fixed64 encoded_name_additional = 12;

  uint32 color_rgba = 13;
  uint64 id = 14;
  uint64 address_in_function = 15;
}

message ApiScopeStopAsync {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;
  uint64 id = 4;
}

message ApiStringEvent {
  // NextID: 15

  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 4;
  fixed64 encoded_name_2 = 5;
  fixed64 encoded_name_3 = 6;
  fixed64 encoded_name_4 = 7;
  fixed64 encoded_name_5 = 8;
  fixed64 encoded_name_6 = 9;
  fixed64 encoded_name_7 = 10;
  fixed64 encoded_name_8 = 11;
  repeated fixed64 encoded_name_additional = 12;

  uint64 id = 13;

  uint32 color_rgba = 14;
}

message ApiTrackInt {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  int32 data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message ApiTrackInt64 {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  int64 data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message ApiTrackUint {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  uint32 data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message ApiTrackUint64 {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  uint64 data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message ApiTrackFloat {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  float data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message ApiTrackDouble {
  // NextID: 15
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;

  double data = 4;

  // Encoded for performance. See `ApiScopeStart` message for details.
  fixed64 encoded_name_1 = 5;
  fixed64 encoded_name_2 = 6;
  fixed64 encoded_name_3 = 7;
  fixed64 encoded_name_4 = 8;
  fixed64 encoded_name_5 = 9;
  fixed64 encoded_name_6 = 10;
  fixed64 encoded_name_7 = 11;
  fixed64 encoded_name_8 = 12;
  repeated fixed64 encoded_name_additional = 13;

  uint32 color_rgba = 14;
}

message Callstack {
  repeated uint64 pcs = 1;

  // This enum describes whether the corresponding stack sample was unwound
  // successfully (kComplete), or the (likely) reason for failure.
  enum CallstackType {
    kComplete = 0;
    kDwarfUnwindingError = 1;
    kFramePointerUnwindingError = 2;
    kInUprobes = 3;
    kInUserSpaceInstrumentation = 7;
    kCallstackPatchingFailed = 4;
    kStackTopForDwarfUnwindingTooSmall = 5;
    kStackTopDwarfUnwindingError = 6;
  }
  CallstackType type = 2;
}

message InternedCallstack {
  uint64 key = 1;
  Callstack intern = 2;
}

message CallstackSample {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 callstack_id = 3;
  uint64 timestamp_ns = 4;
}

message FullCallstackSample {
  uint32 pid = 1;
  uint32 tid = 2;
  Callstack callstack = 3;
  uint64 timestamp_ns = 4;
}

message InternedString {
  uint64 key = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes intern = 2;
}

message InternedTracepointInfo {
  uint64 key = 1;
  TracepointInfo intern = 2;
}

message TracepointEvent {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;
  int32 cpu = 4;
  uint64 tracepoint_info_key = 5;
}

message FullTracepointEvent {
  uint32 pid = 1;
  uint32 tid = 2;
  uint64 timestamp_ns = 3;
  int32 cpu = 4;
  TracepointInfo tracepoint_info = 5;
}

message FullGpuJob {
  uint32 pid = 1;
  uint32 tid = 2;
  uint32 context = 3;
  uint32 seqno = 4;
  int32 depth = 5;
  uint64 amdgpu_cs_ioctl_time_ns = 6;
  uint64 amdgpu_sched_run_job_time_ns = 7;
  uint64 gpu_hardware_start_time_ns = 8;
  uint64 dma_fence_signaled_time_ns = 9;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes timeline = 10;
}

message GpuJob {
  uint32 pid = 1;
  uint32 tid = 2;
  uint32 context = 3;
  uint32 seqno = 4;
  int32 depth = 5;
  uint64 amdgpu_cs_ioctl_time_ns = 6;
  uint64 amdgpu_sched_run_job_time_ns = 7;
  uint64 gpu_hardware_start_time_ns = 8;
  uint64 dma_fence_signaled_time_ns = 9;
  uint64 timeline_key = 10;
}

message GpuQueueSubmission {
  GpuQueueSubmissionMetaInfo meta_info = 1;
  repeated GpuSubmitInfo submit_infos = 2;
  repeated GpuDebugMarker completed_markers = 3;
  // This is the total number of begin markers submitted in this submission.
  int32 num_begin_markers = 4;
}

message GpuQueueSubmissionMetaInfo {
  uint32 tid = 1;
  uint32 pid = 4;
  uint64 pre_submission_cpu_timestamp = 2;
  uint64 post_submission_cpu_timestamp = 3;
}

message GpuSubmitInfo {
  repeated GpuCommandBuffer command_buffers = 1;
}

message GpuCommandBuffer {
  uint64 begin_gpu_timestamp_ns = 1;
  uint64 end_gpu_timestamp_ns = 2;
}

message GpuDebugMarkerBeginInfo {
  // Begin markers can be submitted in a different submission then the end
  // markers, thus the submission meta info of a begin marker might differ from
  // the one of the submission containing the complete debug marker.
  GpuQueueSubmissionMetaInfo meta_info = 1;
  uint64 gpu_timestamp_ns = 2;
}

message GpuDebugMarker {
  GpuDebugMarkerBeginInfo begin_marker = 1;
  uint64 end_gpu_timestamp_ns = 2;
  uint64 text_key = 3;
  int32 depth = 4;
  Color color = 5;
}

message Color {
  // Color ranges are [0.f, 1.f] for the following fields.
  float red = 1;
  float green = 2;
  float blue = 3;
  float alpha = 4;
}

message ThreadName {
  uint32 pid = 1;
  uint32 tid = 2;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes name = 3;
  uint64 timestamp_ns = 4;
}

message ThreadStateSlice {
  reserved 4;
  uint32 pid = 1;  // pid is currently not set as we don't have the information.
  uint32 tid = 2;

  // These are the ones listed in
  // /sys/kernel/debug/tracing/events/sched/sched_switch/format and in
  // https://github.com/torvalds/linux/blob/master/fs/proc/array.c in
  // task_state_array[], with the difference that the OS doesn't distinguish
  // between "running" (scheduled on a CPU) and "runnable" (ready to be
  // scheduled but not actually scheduled), rather they are the same state
  // (often referred to as "running", sometimes as "runnable or running").
  enum ThreadState {
    kRunning = 0;
    kRunnable = 1;
    kInterruptibleSleep = 2;
    kUninterruptibleSleep = 3;
    kStopped = 4;
    kTraced = 5;
    kDead = 6;
    kZombie = 7;
    kParked = 8;
    kIdle = 9;
  }
  ThreadState thread_state = 3;

  uint64 duration_ns = 6;
  uint64 end_timestamp_ns = 5;
}

message AddressInfo {
  uint64 absolute_address = 1;
  uint64 offset_in_function = 2;
  uint64 function_name_key = 3;
  uint64 module_name_key = 4;
}

message FullAddressInfo {
  uint64 absolute_address = 1;
  uint64 offset_in_function = 2;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes function_name = 3;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes module_name = 4;
}

message ModuleUpdateEvent {
  uint32 pid = 1;
  uint64 timestamp_ns = 2;
  ModuleInfo module = 3;
}

message SystemMemoryUsage {
  uint64 timestamp_ns = 1;
  // These fields are retrieved from /proc/meminfo. See
  // https://man7.org/linux/man-pages/man5/proc.5.html for the documentation.
  int64 total_kb = 2;
  int64 free_kb = 3;
  int64 available_kb = 4;
  int64 buffers_kb = 5;
  int64 cached_kb = 6;
  // These fields are retrieved from /proc/vmstat. See
  // https://man7.org/linux/man-pages/man5/proc.5.html for the documentation.
  int64 pgfault = 7;
  int64 pgmajfault = 8;
}

message ProcessMemoryUsage {
  reserved 3;
  uint32 pid = 1;
  uint64 timestamp_ns = 2;
  // These fields are retrieved from /proc/<pid>/stat. See
  // https://man7.org/linux/man-pages/man5/proc.5.html for the documentation.
  int64 minflt = 4;
  int64 majflt = 5;
  // This field is retrieved from /proc/<pid>/status. See
  // https://man7.org/linux/man-pages/man5/proc.5.html for the documentation.
  int64 rss_anon_kb = 6;
}

message CGroupMemoryUsage {
  string cgroup_name = 1;
  uint64 timestamp_ns = 2;
  // This field is retrieved from
  // /sys/fs/cgroup/memory/<cgroup_name>/memory.limit_in_bytes. See
  // https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt for the
  // documentation.
  int64 limit_bytes = 3;
  // These fields are retrieved from
  // /sys/fs/cgroup/memory/<cgroup_name>/memory.stat. See
  // https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt section 5.2
  // for the documentation.
  int64 rss_bytes = 4;
  int64 mapped_file_bytes = 5;
  int64 pgfault = 6;
  int64 pgmajfault = 7;
  int64 unevictable_bytes = 8;
  int64 inactive_anon_bytes = 9;
  int64 active_anon_bytes = 10;
  int64 inactive_file_bytes = 11;
  int64 active_file_bytes = 12;
}

message MemoryUsageEvent {
  uint64 timestamp_ns = 1;
  SystemMemoryUsage system_memory_usage = 2;
  ProcessMemoryUsage process_memory_usage = 3;
  CGroupMemoryUsage cgroup_memory_usage = 4;
}

message ModulesSnapshot {
  uint32 pid = 1;
  uint64 timestamp_ns = 2;
  repeated ModuleInfo modules = 3;
}

message ThreadNamesSnapshot {
  // This is a system-wide snapshot, hence it does not have a pid
  uint64 timestamp_ns = 1;
  repeated ThreadName thread_names = 2;
}

message CaptureFinished {
  enum Status {
    kSuccessful = 0;
    kFailed = 1;
  };

  Status status = 1;
  string error_message = 2;
}

message CaptureStarted {
  // NextID: 9
  uint32 process_id = 1;
  string executable_path = 2;
  string executable_build_id = 3;
  uint64 capture_start_timestamp_ns = 4;
  uint64 capture_start_unix_time_ns = 8;
  uint32 orbit_version_major = 6;
  uint32 orbit_version_minor = 7;
  CaptureOptions capture_options = 5;
}

message WarningEvent {
  uint64 timestamp_ns = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes message = 2;
}

message ErrorEnablingOrbitApiEvent {
  uint64 timestamp_ns = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes message = 2;
}

message ErrorEnablingUserSpaceInstrumentationEvent {
  uint64 timestamp_ns = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes message = 2;
}

message FunctionThatFailedToBeInstrumented {
  uint64 function_id = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  bytes error_message = 2;
}

message WarningInstrumentingWithUserSpaceInstrumentationEvent {
  uint64 timestamp_ns = 1;
  repeated FunctionThatFailedToBeInstrumented
      functions_that_failed_to_instrument = 2;
}

message ClockResolutionEvent {
  uint64 timestamp_ns = 1;
  uint64 clock_resolution_ns = 2;
}

message ErrorsWithPerfEventOpenEvent {
  uint64 timestamp_ns = 1;
  // This is a string, we use bytes to avoid UTF-8 validation.
  repeated bytes failed_to_open = 2;
}

message LostPerfRecordsEvent {
  uint64 duration_ns = 1;
  uint64 end_timestamp_ns = 2;
}

message OutOfOrderEventsDiscardedEvent {
  uint64 duration_ns = 1;
  uint64 end_timestamp_ns = 2;
}

message ClientCaptureEvent {
  reserved 20, 23, 28, 29, 30;

  oneof event {
    // Note that field numbers from 1-15 take only 1 byte to encode
    // https://developers.google.com/protocol-buffers/docs/proto#assigning_field_numbers
    // use them for high frequency events. For the rest please assign
    // numbers starting with 16.
    //
    // Next high-frequency ID: 12
    // Next lower-frequency ID: 49
    // Please keep these alphabetically ordered.

    // Even though AddressInfo is a high-frequency event
    // it is going to go away in the future when we switch to
    // frame-pointer based unwinding.
    AddressInfo address_info = 16;
    ApiEvent api_event = 9;
    ApiScopeStart api_scope_start = 10;
    ApiScopeStartAsync api_scope_start_async = 38;
    ApiScopeStop api_scope_stop = 11;
    ApiScopeStopAsync api_scope_stop_async = 39;
    ApiStringEvent api_string_event = 40;
    ApiTrackDouble api_track_double = 41;
    ApiTrackFloat api_track_float = 42;
    ApiTrackInt api_track_int = 43;
    ApiTrackInt64 api_track_int64 = 44;
    ApiTrackUint api_track_uint = 45;
    ApiTrackUint64 api_track_uint64 = 46;
    CallstackSample callstack_sample = 1;
    CaptureFinished capture_finished = 27;
    CaptureStarted capture_started = 24;
    ClockResolutionEvent clock_resolution_event = 34;
    ErrorEnablingOrbitApiEvent error_enabling_orbit_api_event = 33;
    ErrorEnablingUserSpaceInstrumentationEvent
        error_enabling_user_space_instrumentation_event = 47;
    ErrorsWithPerfEventOpenEvent errors_with_perf_event_open_event = 35;
    FunctionCall function_call = 2;
    GpuJob gpu_job = 3;
    GpuQueueSubmission gpu_queue_submission = 4;
    InternedCallstack interned_callstack = 5;
    InternedString interned_string = 18;
    InternedTracepointInfo interned_tracepoint_info = 19;
    LostPerfRecordsEvent lost_perf_records_event = 36;
    MemoryUsageEvent memory_usage_event = 31;
    ModulesSnapshot modules_snapshot = 25;
    ModuleUpdateEvent module_update_event = 21;
    OutOfOrderEventsDiscardedEvent out_of_order_events_discarded_event = 37;
    SchedulingSlice scheduling_slice = 6;
    ThreadName thread_name = 22;
    ThreadNamesSnapshot thread_names_snapshot = 26;
    ThreadStateSlice thread_state_slice = 7;
    TracepointEvent tracepoint_event = 8;
    WarningEvent warning_event = 32;
    WarningInstrumentingWithUserSpaceInstrumentationEvent
        warning_instrumenting_with_user_space_instrumentation_event = 48;
  }
}

message ProducerCaptureEvent {
  reserved 19, 22, 26, 27, 28;
  oneof event {
    // Note that field numbers from 1-15 take only 1 byte to encode
    // https://developers.google.com/protocol-buffers/docs/proto#assigning_field_numbers
    // use them for high frequency events. For the rest please assign
    // numbers starting with 16.
    //
    // Next high-frequency ID: 15
    // Next lower-frequency ID: 48
    //
    // Please keep these alphabetically ordered.
    ApiEvent api_event = 10;
    ApiScopeStart api_scope_start = 11;
    ApiScopeStartAsync api_scope_start_async = 36;
    ApiScopeStop api_scope_stop = 12;
    ApiScopeStopAsync api_scope_stop_async = 37;
    ApiStringEvent api_string_event = 38;
    ApiTrackDouble api_track_double = 39;
    ApiTrackFloat api_track_float = 40;
    ApiTrackInt api_track_int = 41;
    ApiTrackInt64 api_track_int64 = 42;
    ApiTrackUint api_track_uint = 43;
    ApiTrackUint64 api_track_uint64 = 44;
    CallstackSample callstack_sample = 1;
    CaptureFinished capture_finished = 46;
    CaptureStarted capture_started = 23;
    ClockResolutionEvent clock_resolution_event = 32;
    ErrorEnablingOrbitApiEvent error_enabling_orbit_api_event = 31;
    ErrorEnablingUserSpaceInstrumentationEvent
        error_enabling_user_space_instrumentation_event = 45;
    ErrorsWithPerfEventOpenEvent errors_with_perf_event_open_event = 33;
    FullCallstackSample full_callstack_sample = 2;

    // Even though AddressInfo is a high-frequency event
    // it is going to go away in the future when we switch to
    // frame-pointer based unwinding.
    FullAddressInfo full_address_info = 16;
    FullGpuJob full_gpu_job = 3;
    FullTracepointEvent full_tracepoint_event = 4;
    FunctionCall function_call = 5;
    FunctionEntry function_entry = 13;
    FunctionExit function_exit = 14;
    GpuQueueSubmission gpu_queue_submission = 6;
    InternedCallstack interned_callstack = 7;
    InternedString interned_string = 18;
    LostPerfRecordsEvent lost_perf_records_event = 34;
    MemoryUsageEvent memory_usage_event = 29;
    ModulesSnapshot modules_snapshot = 25;
    ModuleUpdateEvent module_update_event = 20;
    OutOfOrderEventsDiscardedEvent out_of_order_events_discarded_event = 35;
    SchedulingSlice scheduling_slice = 8;
    ThreadName thread_name = 21;
    ThreadNamesSnapshot thread_names_snapshot = 24;
    ThreadStateSlice thread_state_slice = 9;
    WarningEvent warning_event = 30;
    WarningInstrumentingWithUserSpaceInstrumentationEvent
        warning_instrumenting_with_user_space_instrumentation_event = 47;
  }
}
